be/src/util/frame_of_reference_coding.cpp

```diff
@@ -0,0 +1,261 @@
+
+    for (uint32_t i = 0; i < _frame_count + 1; i++ ) {
+        uint32_t offset;
+        _bit_reader.GetValue(ForCoding::INT32_SIZE, &offset);
+        _frame_offsets.push_back(offset);
+    }
+    return true;
+}
+
+template<typename T>
+void ForDecoder<T>::decode_current_frame(T* output) {
+    _current_decoded_frame = _current_index / ForCoding::FRAME_VALUE_NUM;
+    uint8_t frame_value_num = ForCoding::FRAME_VALUE_NUM;
+    if (_current_decoded_frame == _frame_count - 1) {
+        frame_value_num = _last_frame_num;
+    }
+
+    uint32_t base_offset = _frame_offsets[_current_decoded_frame];
+    _bit_reader.SeekToBit(base_offset);
+
+    T min = 0;
+    _bit_reader.GetValue(sizeof(T) * ForCoding::BYTE_SIZE, &min);
+
+    uint32_t frame_bit_size = _frame_offsets[_current_decoded_frame + 1] - base_offset;
+    uint8_t bit_width = (frame_bit_size - sizeof(T) * ForCoding::BYTE_SIZE) / frame_value_num;
+
+    for(uint8_t i = 0; i < frame_value_num; i ++) {
+        T value;
+        _bit_reader.GetValue(bit_width, &value);
+        output[i] = value + min;
+    }
+}
+
+template<typename T>
+T* ForDecoder<T>::copy_value(T* val, size_t count) {
+    memcpy(val, &_out_buffer[_current_index % ForCoding::FRAME_VALUE_NUM], sizeof(T) * count);
+    _current_index += count;
+    val += count;
+    return val;
+}
+
+template<typename T>
+bool ForDecoder<T>::get_batch(T* val, size_t count) {
+    if (_current_index + count > _row_count) {
+        return false;
+    }
+
+    if (need_decode_frame()) {
+        decode_current_frame(_out_buffer);
+    }
+
+    if (_current_index + count < ForCoding::FRAME_VALUE_NUM * (_current_decoded_frame + 1)) {
+        memcpy(val, &_out_buffer[_current_index % ForCoding::FRAME_VALUE_NUM], sizeof(T) * count);
+        _current_index += count;
+        return true;
+    }
+
+    // 1. padding one frame
+    size_t padding_num = ForCoding::FRAME_VALUE_NUM * (_current_decoded_frame + 1) - _current_index;
+    std::cout << "padding_num " << padding_num << std::endl;
+    val = copy_value(val, padding_num);
+
+    // 2. process frame by frame
+    size_t frame_size = (count - padding_num) / ForCoding::FRAME_VALUE_NUM;
+    std::cout << "frame_size " << frame_size << std::endl;
+    for(size_t i = 0; i < frame_size; i++) {
+        // directly decode value to the output, don't  buffer the value
+        decode_current_frame(val);
+        _current_index += ForCoding::FRAME_VALUE_NUM;
+        val += ForCoding::FRAME_VALUE_NUM;
+    }
+
+    // 3. process remaining value
+    size_t remaining_num = (count - padding_num) % ForCoding::FRAME_VALUE_NUM;
+    std::cout << "remaining_num " << remaining_num << std::endl;
+    if (remaining_num > 0) {
+        decode_current_frame(_out_buffer);
+        val = copy_value(val, remaining_num);
+    }
+
+    return true;
+}
+
+template<typename T>
+void ForDecoder<T>::skip(int32_t skip_num) {
+    if (_current_index + skip_num >= _row_count || _current_index + skip_num < 0) {
```

Comment:
Should return false when input is invalid, or that would lead to weird problem.

Reply:
done